src/query/api/v1/httpd/handler.go

```diff
@@ -318,19 +308,91 @@ func (h *Handler) RegisterRoutes() error {
 					Tagged(remoteSource).
 					Tagged(experimentalAPIGroup),
 			)
-			h.router.HandleFunc(annotated.WriteURL,
-				wrapped(experimentalAnnotatedWriteHandler).ServeHTTP,
-			).Methods(annotated.WriteHTTPMethod)
+			addWrappedRoute(annotated.WriteURL, experimentalAnnotatedWriteHandler,
+				annotated.WriteHTTPMethod)
 		}
 	}
 
 	h.registerHealthEndpoints()
 	h.registerProfileEndpoints()
 	h.registerRoutesEndpoint()
 
+	// Register custom endpoints.
+	for _, custom := range h.customHandlers {
+		routeName := routeName(custom.Route(), custom.Methods())
+		route := h.router.Get(routeName)
+		var prevHandler http.Handler
+		if route != nil {
+			prevHandler = route.GetHandler()
+		}
+		customHandler, err := custom.Handler(nativeSourceOpts, prevHandler)
+		if err != nil {
+			return fmt.Errorf("failed to register custom handler with path %s: %w",
+				routeName, err)
+		}
+
+		if route == nil {
+			addWrappedRoute(custom.Route(), customHandler, custom.Methods()...)
+		} else {
+			route.Handler(wrapped(customHandler))
+		}
+	}
+
 	return nil
 }
 
+func addRoute(router *mux.Router, path string, handler http.Handler, methods []string) {
+	addRouteHandlerFn(router, path, handler.ServeHTTP, methods...)
+}
+
+func addRouteHandlerFn(
+	router *mux.Router,
+	path string,
+	handlerFn func(http.ResponseWriter, *http.Request),
+	methods ...string,
+) {
+	routeName := routeName(path, methods)
+	if previousRoute := router.Get(routeName); previousRoute != nil {
+		// route already exists.
+		return
+	}
+
+	route := router.
+		HandleFunc(path, handlerFn).
+		Name(routeName)
+
+	if len(methods) > 0 {
+		route.Methods(methods...)
+	}
+}
+
+func routeName(p string, methods []string) string {
+	if len(methods) > 1 {
+		distinctMethods := distinct(methods)
+		sort.Strings(distinctMethods)
+		return path.Join(p, strings.Join(distinctMethods, ":"))
+	}
+
+	return path.Join(p, strings.Join(methods, ":"))
+}
+
+func distinct(arr []string) []string {
+	if len(arr) <= 1 {
+		return arr
+	}
+
+	foundEntries := make(map[string]bool, len(arr))
+	result := make([]string, 0, len(arr))
+	for _, entry := range arr {
+		if ok := foundEntries[entry]; !ok {
+			foundEntries[entry] = true
+			result = append(result, entry)
+		}
+	}
```

Comment:
nit: these are probably going to be quite small (max of 5 elements); may be better to just do a slice scan and not use the map at all. Considering these should be sorted anyway can even do it in place with something like this:

```golang
var prev string
distinct := arr[:0]
for i, entry := range arr {
 if i == 0 || prev != entry {
  distinct = append(distinct, entry)
 } 
}

return distinct
```